### PR TITLE
Add updates count badge to Updates sidebar item

### PR DESCRIPTION
**Pull request by janfaracik**
Screenshot 2022-09-08 at 21 21 57

Adds back the updates count badge introduced and removed in #6783, this time by adding an attribute to the `task.jelly` component so that any view/plugin can use it.

### Proposed changelog entries

* Add updates count badge to Updates sidebar item.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  - Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content-Security-Policy directives (see [documentation on jenkins.io](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@jenkinsci/sig-ux 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7084"></a>

### CODE REVIEW

**Review by timja — APPROVED** [review]
lgtm, tested locally

image

image

image

File: core/src/main/resources/lib/layout/task.jelly
Comment (NotMyFault) [inline core/src/main/resources/lib/layout/task.jelly:82]:
```suggestion
      (@since TODO)
```
We can use an @ here for sure.

**Review by NotMyFault — APPROVED** [review]
Looks nice :)

(feedback is non-blocking)

File: core/src/main/resources/lib/layout/task.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/task.jelly:144]: This seems inconsistent with the naming convention you're generally using, why?

Comment (janfaracik) [reply core/src/main/resources/lib/layout/task.jelly:144]: I'm just following the existing naming convention for this component - I'd be happy to rename the classes (and a general cleanup of the component) but that'd be out of scope for this PR.

File: core/src/main/resources/lib/layout/task.jelly
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/task.jelly:81]:
```suggestion
      If set, displays the value as a small badge on the right side of the sidepanel item.
```

**Review by daniel-beck — DISMISSED** [review]
Looks reasonable. Implementation is fine.

There should be usage guidelines.

**Comment by timja** [comment]
> There should be usage guidelines.

@janfaracik is that something you're able to add to design library before we ship this?

**Comment by janfaracik** [comment]
> > There should be usage guidelines.
> 
> 
> 
> @janfaracik is that something you're able to add to design library before we ship this?

Will do, am I good to add it to the existing 'Layouts' PR?

**Comment by daniel-beck** [comment]
I think there should there be support for badges in context menus. For a generalized feature (e.g. imagine badges with open issues in various `warnings-ng` contributed sidepanel tasks) it probably makes sense to support that.

**Comment by timja** [comment]
> Will do, am I good to add it to the existing 'Layouts' PR?

Yes 👍

**Comment by janfaracik** [comment]
> > Will do, am I good to add it to the existing 'Layouts' PR?
> 
> Yes 👍

Updated that PR with some examples of when to use badges:

image

**Comment by timja** [comment]
Do you plan to address Daniel's comment on badges in context menu's in this PR?

**Comment by janfaracik** [comment]
> Do you plan to address Daniel's comment on badges in context menu's in this PR?

I'd favour doing them separately to this PR (just to get this in as I think its a nice-to-have), happy to do it though in this if you prefer @daniel-beck?

**Comment by daniel-beck** [comment]
I'd prefer that this is done at the same time, to have parity between sidepanel and menus. (Unless there's a UX reason that menus should not have these badges, now or later, of course.)

**Comment by timja** [comment]
@janfaracik do you plan to take a look at the context menu changes?

**Comment by janfaracik** [comment]
> @janfaracik do you plan to take a look at the context menu changes?

I will do for sure, just a little swamped at the moment.

**Comment by janfaracik** [comment]
Updated the PR so that dropdown menus now support badges, as well as items in the Manage Jenkins page.

image

---

Currently badges just support numbers, could there be a use case for returning text?

File: core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java
Comment (timja) [inline core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java:340]: this should still be private just add a getter for stapler, historical methods have had the suppression added

**Comment by uhafner** [comment]
> Currently badges just support numbers, could there be a use case for returning text?

This is a good question. I think that we should not restrict it to only numbers in order to prevent future API breakages. Most badges support at least percentages (e.g., 83%).

**Comment by uhafner** [comment]
> > > Currently badges just support numbers, could there be a use case for returning text?
> > 
> > 
> > This is a good question. I think that we should not restrict it to only numbers in order to prevent future API breakages. Most badges support at least percentages (e.g., 83%).
> 
> see discussion here as well: [#6995 (comment)](https://github.com/jenkinsci/jenkins/pull/6995#discussion_r945711898)

That means it would be better to use a string for the content. I'm not sure if it makes sense to define an upper limit for the number of visible characters, I hope that developers will use that feature wisely.

**Comment by timja** [comment]
I've merged this feature with https://github.com/jenkinsci/jenkins/pull/6995 in https://github.com/janfaracik/jenkins/compare/add-updates-badge...timja:jenkins:add-updates-badge-2?expand=1

All working, just missing tooltips support which came from the other feature which shouldn't be hard to add

File: core/src/main/java/hudson/model/UpdateCenter.java
Comment (timja) [inline core/src/main/java/hudson/model/UpdateCenter.java:382]: code is just moved from `PluginsLink`

**Review by timja — APPROVED** [review]
I've:

- resolved the conflicts with https://github.com/jenkinsci/jenkins/pull/6995, 
- added tooltips to badges on sidebar and context menu which that PR added for management links
- changed from number to badge class so strings now work although should follow the guidelines in the badge class:

https://github.com/jenkinsci/jenkins/blob/a6d542509799d1b70610406c7750f2440c44e8b2/core/src/main/java/jenkins/management/Badge.java#L60-L62

Re-tested all locally by checking with old data and new plugin updates available

File: core/src/main/resources/lib/layout/breadcrumbs.js
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/breadcrumbs.js:28]: Unspecified parameters are `undefined` though? Or what is this case for?

File: core/src/main/resources/lib/layout/breadcrumbs.js
Comment (daniel-beck) [inline core/src/main/resources/lib/layout/breadcrumbs.js:30]: Looks XSS-able. Badge tooltips are fairly likely to contain user-provided content. Please create the element via DOM API (or at least not via string concatenation / template string).

File: core/src/main/resources/lib/layout/breadcrumbs.js
Comment (daniel-beck) [reply core/src/main/resources/lib/layout/breadcrumbs.js:28]: Nvm, it's when an explicit `null` is provided.

File: core/src/main/resources/lib/layout/breadcrumbs.js
Comment (daniel-beck) [reply core/src/main/resources/lib/layout/breadcrumbs.js:30]: XSS

**Review by daniel-beck — DISMISSED** [review]

File: core/src/main/resources/lib/layout/breadcrumbs.js
Comment (timja) [reply core/src/main/resources/lib/layout/breadcrumbs.js:30]: could you provide steps to reproduce please?

Comment (daniel-beck) [reply core/src/main/resources/lib/layout/breadcrumbs.js:30]: Write a plugin that declares a `ManagementLink` with user-defined `Badge` title and tooltip.

Use the XSS payloads `` for the `title` and/or `">` for the `tooltip`.

I don't exactly remember what values I used, but it should be straightforward enough based on this code.

Comment (janfaracik) [reply core/src/main/resources/lib/layout/breadcrumbs.js:30]: Should any further changes to this/menu logic be delayed until #7474 is in?

Comment (timja) [reply core/src/main/resources/lib/layout/breadcrumbs.js:30]: we could fix this error and then move on if the other one will take longer?

Comment (janfaracik) [reply core/src/main/resources/lib/layout/breadcrumbs.js]: Fixed this, would you mind retesting when you get the chance? Thanks

Comment (timja) [reply core/src/main/resources/lib/layout/breadcrumbs.js]: @daniel-beck  / @jenkinsci/core-security-review do you mind re-checking this please?

**Review by NotMyFault — APPROVED** [review]
The number in the plugin manager could use the same color as the badge in the manage view:
Screenshot 2023-03-25 at 12 58 08
But I consider this non-blocking feedback, thanks for your work!

**Review by daniel-beck — DISMISSED** [review]
The tooltip isn't just in a "plain" HTML environment, but in an HTML attribute.

A tooltip of `" onmouseenter=alert(1) "` causes XSS.

As a general recommendation, stop concatenating strings to create HTML entirely.

File: core/src/main/resources/lib/layout/breadcrumbs.js
Comment (daniel-beck) [reply core/src/main/resources/lib/layout/breadcrumbs.js]: Still broken per https://github.com/jenkinsci/jenkins/pull/7084#pullrequestreview-1360970874 because my advice was ignored.

Comment (janfaracik) [reply core/src/main/resources/lib/layout/breadcrumbs.js]: Good spot, I've updated it using xmlEscape from 'security/util.js'. Let me know if that's okay with you, the scenario you outlined below no longer occurs.

**Comment by janfaracik** [comment]
Is this PR good for merge notice now @timja?

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

File: core/src/main/resources/lib/layout/task.jelly
Comment (mawinter69) [inline core/src/main/resources/lib/layout/task.jelly:79]: This needs a better description. At the moment most people will assume this is just a simple text you can pass in but what is actually is required is a jenkins.management.Badge object.

**Pull request merged**
